ambry-network/src/main/java/com/github/ambry/network/SocketNetworkClient.java

```diff
@@ -158,14 +156,19 @@ public SocketNetworkClient(Selector selector, NetworkConfig networkConfig, Netwo
       if (time.milliseconds() - requestMetadata.requestQueuedAtMs > checkoutTimeoutMs) {
         responseInfoList.add(
             new ResponseInfo(requestMetadata.requestInfo, NetworkClientErrorCode.ConnectionUnavailable, null));
+        requestMetadata.requestInfo.getRequest().release();
         logger.trace("Failing request to host {} port {} due to connection unavailability",
             requestMetadata.requestInfo.getHost(), requestMetadata.requestInfo.getPort());
         iter.remove();
-        requestMetadata.requestInfo.getRequest().release();
         if (requestMetadata.pendingConnectionId != null) {
           pendingConnectionsToAssociatedRequests.remove(requestMetadata.pendingConnectionId);
           requestMetadata.pendingConnectionId = null;
         }
+        String connId =
+            correlationIdInFlightToConnectionId.get(requestMetadata.requestInfo.getRequest().getCorrelationId());
+        if (connId != null) {
+          connectionIdToRequestInFlight.remove(connId);
```

Comment:
I suppose we should checkin the connection here. There is a chance that when a request is associated with a connection and somehow it takes so long to send it out that this request just times out. In this case, we are releasing the request's bytebuf here in this line 159, so we have to make sure the request is removed from all bookkeeping maps and sets.

Reply:
I guess this piece of code doesn't hurt. But I still don't see how `correlationIdInFlightToConnectionId.get` will ever return non-null at this point. This part of the code handles requests that do not yet have a connection to make the request over, and that map is only added to after a connection is checked out. 

One confusing part of `SocketNetworkClient` is how the pending connection ID works. The pending connection ID is not actually checked out yet but is an indicator to show that a new connection is in the process of being established and will eventually be added to the pool but weirdly it doesn't guarantee that that specific connection is actually used for the request it was associated with.

Reply:
I think we have a disagreement on whether a request will timeout after it gets associated with a connection id. I certainly think it's possible that even after a request has checked out a connection, it can still time out, due to slow outbound link, gc or some other reasons. 

In MockSelector of unit tests, we do have a case where the MockSelector would throw IOException when sending out a request without disconnecting the connection. SocketNetworkClient won't remove this request from pending requests when seeing an IOException. So I take that we will retry this request, on the same connection, until it times out.